### DISCUSSION TITLE
dependency use yield doesn't response to the client immediately after yield, still have to wait after all the code after yield has finished

### DISCUSSION THREAD

**Discussion by annexationer** [fastapi/fastapi#10004] [Questions] [answered]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

I have a dependency like this:
![image](https://github.com/tiangolo/fastapi/assets/91235829/d4575d62-e558-4bfb-bc62-811e06b192b4)
I use await asyncio.sleep(20) after yield, I expect the response return to the client immediately, after 20 seconds, there will be a print 'finally', but I test with curl, the response is immediate, but in the postman or browser, I still have to wait 20 seconds, I thought it was because connection header, but I changed the connection header, still the same result.
<img width="2038" alt="image" src="https://github.com/tiangolo/fastapi/assets/91235829/98a16023-3112-4605-a820-e2c39456bd9c">
<img width="426" alt="image" src="https://github.com/tiangolo/fastapi/assets/91235829/525fa2c1-e4fc-44c2-9739-dec46ee76d1d">

I just really want to know why, is this a bug?

**Comment by sulavmhrzn** [comment]
direct use of `asyncio.sleep()` can potentially block the event loop. Use BackgroundTask if you really need to wait 20 seconds. 
When you run a curl command, it won't wait for the asynchronous sleep to complete, and it will return immediately. However, if you are making an actual HTTP request to your FastAPI server from a client, the client should wait for the server's response.

**Comment by mbbyn** [comment]
I would also really like to know the reason for this behavior. Using `BackgroundTask` or performing work after the `yield` should be equivalent, since [both run after the response is returned][seq-diagram]. It feels like some clients terminate eagerly once the body chunks are sent, while others wait until the server terminates the connection? I am really not sure, but would be very interested to know as this issue makes `BackgroundTasks` risky if their behavior is not clearly understood.

Also as a reference, you can see how Starlette is running background tasks after sending the body:

https://github.com/encode/starlette/blob/93074e82ab15e860456d4a7074c0494eed78ea0c/starlette/responses.py#L145-L156

[seq-diagram]: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-httpexception

**Comment by YuriiMotov** [comment, marked as answer]
This is probably done because in some cases you may want to return server error instead of normal response in case of any errors happen during the execution of code after `yield`:

```
import asyncio
from fastapi import Depends, FastAPI, HTTPException

app = FastAPI()

async def my_dep():
    yield
    await asyncio.sleep(5)
    if True:  # Some condition
        raise HTTPException(status_code=500, detail="Error after yield")

@app.get("/app")
def read_main(d: None = Depends(my_dep)):
    return {"message": "Hello World"}
```

**BTW**, seems that in current version of FastAPI (0.115.12), `curl` behaves the same result as browser:
```
time curl --location 'http://localhost:8000/app'
{"detail":"Error after yield"}
real	0m5.009s
user	0m0.002s
sys	0m0.006s

```

**Reply by mbbyn** [reply]
Thanks for this, it justifies the behavior very well.

Do you know an alternative for scheduling background tasks that don't interfere with the request lifecycle, besides manually building a `asyncio.Task` queue externally and managing it manually using lifespan?

**Reply by YuriiMotov** [reply]
It depends on your requirements.
You would better open a new question to ask this
